alethzero/plugins/whisper/Whisper.cpp

```diff
@@ -175,70 +207,87 @@ void Whisper::on_newIdentity_triggered()
 	noteTopic(qsTopic);
 
-	QString const qsDest = m_ui->shhTo->currentText();
-	Public dest = stringToPublic(qsDest);
+	QString const strDest = m_ui->shhTo->currentText();
+	Public dest = stringToPublic(strDest);
 	if (dest)
-		noteDestination(qsDest);
-	else if (!qsDest.isEmpty())
-		return; // don't allow unencrypted messages, unless it was explicitly intended
+	{
+		if (m_ui->shhTo->findText(strDest) < 0)
+			m_ui->shhTo->addItem(strDest);
+	}
+	else if (!strDest.isEmpty())
+	{
+		// don't allow unencrypted messages, unless it was explicitly intended
+		QMessageBox box(QMessageBox::Warning, "Warning", "Invalid destination address!");
+		box.setInformativeText("Destination address must be either valid or blank.");
+		box.exec();
+		return;
+	}
 
-	string text = m_ui->shhData->toPlainText().toStdString();
-	if (text.empty())
+	int const msgSizeLimit = 1024;
+	QString text = m_ui->shhData->toPlainText();
+	if (text.isEmpty())
 		return;
+	else if (text.size() > msgSizeLimit)
+	{
+		QMessageBox box(QMessageBox::Warning, "Warning", "Message too big!");
+		box.setInformativeText(QString("Single message should not exceed %1 characters.").arg(msgSizeLimit));
+		box.exec();
+		return;
+	}
 
 	shh::Message m;
 	m.setTo(dest);
-	m.setPayload(asBytes(text));
+	m.setPayload(asBytes(text.toStdString()));
 
 	Secret from;
 	Public f = stringToPublic(m_ui->shhFrom->currentText());
 	if (zero()->web3Server()->ids().count(f))
 		from = zero()->web3Server()->ids().at(f);
 
-	shh::BuildTopic topic(qsTopic.toStdString());
+	shh::Topics topic = stringToTopics(qsTopic);
 	web3()->whisper()->inject(m.seal(from, topic, m_ui->shhTtl->value(), m_ui->shhWork->value()));
 	m_ui->shhData->clear();
 	m_ui->shhData->setFocus();
 }
 
-void Whisper::on_forget_clicked()
+void Whisper::onForgetDestinationsClicked()
 {
-	if (m_knownTopics.empty())
-		return;
+	m_ui->shhTo->clear();
+}
 
-	m_knownTopics.clear();
+void Whisper::onForgetTopicsClicked()
+{
 	m_ui->shhTopic->clear();
-
-	auto x = zero()->findPlugin(c_chatPluginName);
-	WhisperPeers* wp = dynamic_cast<WhisperPeers*>(x.get());
-	if (wp)
-		wp->forgetTopics();
 }
 
-void Whisper::noteTopic(QString const _topic)
+void Whisper::noteTopic(QString const& _topic)
 {
-	if (_topic.isEmpty() || m_knownTopics.contains(_topic))
+	if (_topic.isEmpty())
 		return;
 
-	m_knownTopics.insert(_topic);
-	m_ui->shhTopic->addItem(_topic);
-
 	auto x = zero()->findPlugin(c_chatPluginName);
 	WhisperPeers* wp = dynamic_cast<WhisperPeers*>(x.get());
-	if (wp)
-		wp->noteTopic(_topic);
-}
+	if (!wp)
+		return;
 
-void Whisper::noteDestination(QString const _dest)
-{
-	if (!m_destinations.contains(_dest))
+	QStringList tx = _topic.split("|", QString::SkipEmptyParts);
+
+	if (tx.size() > 1 && m_ui->shhTopic->findText(_topic) < 0)
+		m_ui->shhTopic->addItem(_topic);
+
+	for (auto t: tx)
 	{
-		m_destinations.insert(_dest);
-		m_ui->shhTo->addItem(_dest);
+		if (t.isEmpty())
+			continue;
+
+		if (m_ui->shhTopic->findText(t) < 0)
```

Comment:
This line is same as line 280 if topics.size() is 2. Should line 280 check if topcis.size() ==1, instead? (and this body of code can be placed under else)